File: Algorithms/golang-sorting/sort/mergesort/mergesortOptimized.go
package mergesort

import "heartiger.com/algorithms/sort/insertionsort"


// Merge sort recursively split the array in half
// Until the array with only 1 element, it sorted.
// Then start to merge two array into one.
// So start with split array down then merge array up recursively.
// // Time Complexity O(nlogn)
func MergeSorterOptimized(nums []int){
	mergeSortOptimized(nums, 0, len(nums)-1)
}

func mergeSortOptimized(nums []int, left int, right int) {

	// (Optimization): cutoff set as 15 to use insertion sort
	if left - right <= 15 {
		insertionsort.InsertionSortWithIndexes(nums, left, right+1)
		return
	}

	mid := left + (right - left) / 2
	mergeSortOptimized(nums, left, mid)
	mergeSortOptimized(nums, mid+1, right)

	// (Optimization): if the last element of first array already less than the first element
	// Of the second array. It's sorted array no need to merge.
	// It's efficient on almost sorted array. Also performance reduction as side affect.
	if nums[mid] > nums[mid+1] {
		mergeOptimized(nums, left, mid, right)
	}
}
func mergeOptimized(nums []int, left int, mid int, right int) {

	// Create two array(slice) to hold values to be merged
	sub1Len := mid - left + 1
	sub2Len := right - mid

	sub1 := make([]int, sub1Len)
	sub2 := make([]int, sub2Len)

	for i:=0; i<sub1Len; i++ {
		sub1[i] = nums[left + i]
	}
	// Since second array start at mid+1
	for i:=0; i<sub2Len; i++ {
		sub2[i] = nums[mid + 1 + i]
	}

	index := left
	i := 0
	j := 0

	// Merge values in asc order
	for i < sub1Len && j < sub2Len {
		if sub1[i] < sub2[j] {
			nums[index] = sub1[i]
			i++
			index++
		} else {
			nums[index] = sub2[j]
			j++
			index++
		}
	}

	// Merge the remaining values
	for i < sub1Len {
		nums[index] = sub1[i]
		i++
		index++
	}
	for j < sub2Len {
		nums[index] = sub2[j]
		j++
		index++
	}
}

File: Algorithms/Sorting/src/sorters/MergeSortOptimized.java
package sorters;

import utility.Sortable;

public class MergeSortOptimized implements Sortable {
    private int[] nums;

    public MergeSortOptimized(int[] arr){
        nums = arr;
    }
    public MergeSortOptimized(){}

    @Override
    public int[] sort(){
        sortArray(0, nums.length-1);
        return nums;
    }

    @Override
    public void setArray(int[] nums) {
        this.nums = nums;
    }

    private void sortArray(int left, int right){
        if(right - left <= 15){
            InsertionSort.sort(nums, left, right);
            return;
        }
        int mid = left + (right - left)/2;
        sortArray(left, mid);
        sortArray(mid+1, right);
        // If left side array's right most element is lesser than and equal
        // to right side array's left most element, no need to merge.
        if(nums[mid] > nums[mid+1])
            mergeArray(left, mid, right);
    }

    private void mergeArray(int left, int mid, int right) {
        int sizeA = mid - left + 1;
        int sizeB = right - mid;

        int[] subArrayA = new int[sizeA];
        for(int a=0; a<sizeA; a++)
            subArrayA[a] = nums[left + a];

        int[] subArrayB = new int[sizeB];
        for(int b=0; b<sizeB; b++)
            subArrayB[b] = nums[mid + b + 1];

        int a = 0;
        int b = 0;
        int index = left;
        while(a < sizeA && b < sizeB){
            if(subArrayA[a] < subArrayB[b])
                nums[index++] = subArrayA[a++];
            else
                nums[index++] = subArrayB[b++];
        }

        while(a < sizeA)
            nums[index++] = subArrayA[a++];

        while(b < sizeB)
            nums[index++] = subArrayB[b++];
    }


}
